concordium-consensus/src/Concordium/GlobalState/Persistent/BlockState/ProtocolLevelTokens.hs

```diff
@@ -0,0 +1,324 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DerivingVia #-}
+{-# LANGUAGE TypeApplications #-}
+
+module Concordium.GlobalState.Persistent.BlockState.ProtocolLevelTokens where
+
+import Control.Monad
+import Data.Bits
+import qualified Data.ByteString as BS
+import qualified Data.ByteString.Short as SBS
+import qualified Data.Map.Strict as Map
+import Data.Serialize
+import Data.Word
+
+import qualified Concordium.Crypto.SHA256 as SHA256
+import Concordium.Types.HashableTo
+import Concordium.Types.ProtocolVersion
+import Concordium.Utils
+import Concordium.Utils.Serialization
+
+import Concordium.GlobalState.Basic.BlockState.LFMBTree (LFMBTreeHash' (..))
+import Concordium.GlobalState.Persistent.BlobStore
+import qualified Concordium.GlobalState.Persistent.LFMBTree as LFMBTree
+
+-- | Represents the raw amount of a token. This is the amount of the token in its smallest unit.
+newtype TokenRawAmount = TokenRawAmount {theTokenRawAmount :: Word64}
+    deriving newtype (Eq, Ord, Show, Num, Real)
+
+-- | Serialization of 'TokenRawAmount' is as a variable length quantity (VLQ). We disallow
+--  0-padding to enforce canonical serialization.
+instance Serialize TokenRawAmount where
+    put (TokenRawAmount amt) = do
+        mapM_ putWord8 (chunk amt [])
+      where
+        chunk num []
+            | num == 0 = [0]
+            | otherwise = chunk (num `shiftR` 7) [fromIntegral $ num .&. 0x7f]
+        chunk num l
+            | num == 0 = l
+            | otherwise = chunk (num `shiftR` 7) (fromIntegral (0x80 .|. (num .&. 0x7f)) : l)
+    get = TokenRawAmount <$> loop 0
+      where
+        loop accum = do
+            b <- getWord8
+            when (b == 0x80 && accum == 0) $
+                fail "Padding bytes are not allowed"
+            when (accum > maxBound `shiftR` 7) $
+                fail "Value out of range"
+            let accum' = accum `shiftL` 7 .|. fromIntegral (b .&. 0x7f)
+            if testBit b 7
+                then loop accum'
+                else return accum'
+
+instance (MonadBlobStore m) => BlobStorable m TokenRawAmount
+
+-- FIXME: Refactor to base.
```

Comment:
Do you want to have the types below in base, or here? In https://github.com/Concordium/concordium-base/pull/612#event-17116565552 you mention this PR.

Reply:
`TokenId` has moved. I think `TokenModuleRef` will move, but I will leave it for this PR. `TokenIndex` I think will not move to base, since it does not need to be exposed and is just an internal implementation detail.